### PR TITLE
Code Review

### PR DESCRIPTION
**Pull request by atsheehan**

### CODE REVIEW

File: app/controllers/cars_controller.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  end
+
+  def new
+    @car = Car.new
+    @manufacturers = Manufacturer.all
+  end
+
+  def create
+    @car = Car.create(car_params)
+    @manufacturers = Manufacturer.all
```
Comment (atsheehan) [inline app/controllers/cars_controller.rb:13]: You actually only need to query for the manufacturers if you're going to render the `new` template. You can move this into the `else` block so you save yourself a query when the car is saved correctly:

``` ruby
if @car.save
  # Can avoid having to re-query for manufacturers here.
  redirect_to cars_path
else
  @manufacturers = Manufacturer.all
  render :new
end
```

File: app/controllers/manufacturers_controller.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    @manufacturers = Manufacturer.all
+  end
+
+  def new
+    @manufacturer = Manufacturer.new
+  end
+
+  def create
+    @manufacturer = Manufacturer.new(manufacturer_params)
+    @manufacturer.save
```
Comment (atsheehan) [inline app/controllers/manufacturers_controller.rb:13]: Extra call to `save` here is redundant.

File: app/models/car.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  belongs_to :manufacturer
+
+  validates :color, presence: true
+  validates :year, presence: true
+  validates :mileage, presence: true
+  validates_numericality_of :mileage, :greater_than_or_equal_to => 0
+  validates :model_name, presence: true
+  validates :year, presence: true
+  validates_numericality_of :year, :greater_than_or_equal_to => 1920
+  validates_numericality_of :manufacturer_id
```
Comment (atsheehan) [inline app/models/car.rb:11]: Rather than validating the numericality of the ID value, you can validate the presence of an actual manufacturer record:

``` ruby
validates :manufacturer, presence: true
```

This checks that it contains a valid integer and that the ID points to a record in the manufacturers table.

File: app/models/car.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+class Car < ActiveRecord::Base
+  belongs_to :manufacturer
+
+  validates :color, presence: true
+  validates :year, presence: true
+  validates :mileage, presence: true
+  validates_numericality_of :mileage, :greater_than_or_equal_to => 0
+  validates :model_name, presence: true
+  validates :year, presence: true
```
Comment (atsheehan) [inline app/models/car.rb:9]: Duplicate.

File: app/models/car.rb [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+  validates_numericality_of :year, :greater_than_or_equal_to => 1920
```
Comment (atsheehan) [inline app/models/car.rb:10]: You can combine several validations for a single attribute:

``` ruby
validates :year, presence: true, numericality: { greater_than_or_equal_to: 1920 }
```

File: app/models/manufacturer.rb [hunk 6]
```diff
@@ -0,0 +1,6 @@
+class Manufacturer < ActiveRecord::Base
+  has_many :cars
+
+  validates :name, presence: true
```
Comment (atsheehan) [inline app/models/manufacturer.rb:4]: You might want to validate the uniqueness of the manufacturer name:

``` ruby
validates :name, presence: true, uniqueness: true
```

File: app/views/cars/index.html.erb [hunk 7]
```diff
@@ -0,0 +1,15 @@
+<h1>Cars, Currently in Inventory:</h1>
+  <h3><em>Sell 'em while they're hot!</em></h3>
+
+<ul>
+  <% if @cars.count == 0 %>
+    <li><h3><a href='/cars/new'>Add the first car!</h3></li>
```
Comment (atsheehan) [inline app/views/cars/index.html.erb:6]: You can use the view helpers to simplify creating links:

``` HTML+ERB
<%= link_to 'Add the first car!', new_car_path %>
```

File: app/views/cars/new.html.erb [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+      </div>
+
+      <div>
+        <%= f.label :mileage %>
+        <%= f.text_field :mileage %>
+      </div>
+
+      <div>
+        <%= f.label :manufacturer_id %>
+        <%= f.select(:manufacturer_id, Manufacturer.all.collect {|p| [ p.name, p.id ] }, {include_blank: 'Select a Manufacturer'}) %>
```
Comment (atsheehan) [inline app/views/cars/new.html.erb:44]: You've already queried for the manufacturers within your controller and you're querying again here. Instead you can use the `@manufacturers` instance variable you set:

``` ruby
@manufacturers.collect { |m| [m.name, m.id] }
```

File: car_sales/.gitignore [hunk 9]
```diff
@@ -1,16 +0,0 @@
...
-# Ignore bundler config.
-/.bundle
-
-# Ignore the default SQLite database.
-/db/*.sqlite3
-/db/*.sqlite3-journal
-
-# Ignore all logfiles and tempfiles.
-/log/*.log
-/tmp
```
Comment (atsheehan) [inline car_sales/.gitignore:16]: When moving the files from the `car_sales` directory to the root you forgot to bring the `.gitignore` file. This is important to have so that you don't commit lots of useless log and tmp files to your repository.

File: config/routes.rb [hunk 10]
```diff
@@ -1,4 +1,7 @@
 Rails.application.routes.draw do
+  root 'welcome#index'
+  resources :cars, only: [:index, :new, :create]
+  resources :manufacturers, only: [:index, :new, :create]
```
Comment (atsheehan) [inline config/routes.rb:4]: You might want to include a `show` page for the manufacturers and cars so that you can view the details of each.

File: db/migrate/20140628150243_newtables.rb [hunk 11]
```diff
@@ -0,0 +1,19 @@
+class Newtables < ActiveRecord::Migration
```
Comment (atsheehan) [inline db/migrate/20140628150243_newtables.rb:1]: I would use a more descriptive name for the migration (e.g. `CreateCarsAndManufacturers`)

File: spec/features/car_spec.rb [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+
+    attrs_car = {
+      model_name: 'S 65 AMG',
+      year: '2015',
+      color: 'Black',
+      mileage: 15,
+      manufacturer_id: manufacturer.id
+    }
+
+    car = Car.create!(attrs_car)
```
Comment (atsheehan) [inline spec/features/car_spec.rb:35]: You can use FactoryGirl to simplify creating some of these objects.

File: spec/features/car_spec.rb [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+    fill_in 'Model name', with: 'SLS AMG'
+    fill_in 'Year', with: 2015
+    fill_in 'Color', with: 'Red'
+    fill_in 'Mileage', with: 30
+    fill_in 'Description', with: 'Incredibly fast.'
+    click_button 'Submit'
+    expect(page).to have_content 'Manufacturer is not a number'
+    expect(page).to have_content 'Something went wrong'
+  end
+end
```
Comment (atsheehan) [inline spec/features/car_spec.rb:195]: For acceptance tests you don't have to enumerate every possible edge case. What you really want to test is that the form is properly integrated and that you handle the success path (everything checks out, a new car is created and the user is redirected) as well as the error path (something on the form is invalid, redisplay the form with error messages).

For more fine-grained control of the validations I would use model tests which are faster to create and run.
